docs: initial storage cleanup (#2433)
AztecProtocol/aztec-packages: docs/docs/dev_docs/contracts/syntax/storage.md

```rust
impl Storage {
  fn init(context: Context) -> Self {
    Storage {
      // (public state variables)::new()
      // (private state variables)::new()
    }
  }
}
```

To use storage in functions, e.g., functions where you would read or write storage, you need to initialize the struct first, and then you can read and write afterwards. Below are snippets for initializing in private and public functions.

#include_code private_init_storage /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust
#include_code public_init_storage /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

:::info
https://github.com/AztecProtocol/aztec-packages/pull/2406 is removing the need to explicitly initialize the storage in each function before reading or writing. This will be updated in the docs once the PR is merged.
:::

:::warning Using slot `0` is not supported!
No storage values should be initialized at slot `0`. This is a known issue that will be fixed in the future.
:::

## Map

> A `map` is a state variable that "maps" a key to a value. In Aztec.nr, we keys are `Field`s (or values that are convertible to Fields) and values can be any type - even other maps. The map is a struct defined as follows:

Reviewer: ```suggestion
A `map` is a state variable that "maps" a key to a value. In Aztec.nr, keys are `Field`s (or values that are convertible to Fields) and values can be any type - even other maps. The map is a struct defined as follows:
```

#include_code map /yarn-project/aztec-nr/aztec/src/state_vars/map.nr rust

> It includes a [`Context`](./context.mdx) to be used when dabbling in the private domain, a `storage_lot` to specify where in storage the map is stored, and a `start_var_constructor` which tells the map how it should operate on the underlying type. This includes how to serialize and deserialize the type, as well as how commitments and nullifiers are computed for the type if private.

Reviewer: ```suggestion
It includes a [`Context`](./context.mdx) to be used when dabbling in the private domain, a `storage_slot` to specify where in storage the map is stored, and a `start_var_constructor` which tells the map how it should operate on the underlying type. This includes how to serialize and deserialize the type, as well as how commitments and nullifiers are computed for the type if it's private.
```

#include_code new /yarn-project/aztec-nr/aztec/src/state_vars/map.nr rust

We will see examples of map constructors for public and private variables in later sections.

### `at`

When dealing with a map, we can access the value at a given key using the `::at` method. This takes the key as an argument and returns the value at that key.

#include_code at /yarn-project/aztec-nr/aztec/src/state_vars/map.nr rust

This function behaves similarly for both private and public maps. An example could be if we have a map with `minters`, which is mapping addresses to a flag for whether they are allowed to mint tokens or not.

#include_code read_minter /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

Above, we are specifying that we want to get the storage in the Map `at` the `msg_sender()`, read the value stored and check that `msg_sender()` is indeed a minter. Doing a similar operation in Solidity code would look like:

```solidity
require(minters[msg.sender] == 1, "caller is not minter");
```

## Public State Variables

> As we have looked seen earlier we got two kinds of state, public and private. We'll start with a look at the public side of things, as this is what behaves closest to what most people are used to from Ethereum and other blockchains.

Reviewer: ```suggestion
As we have seen earlier we have two kinds of state, public and private. We'll start with a look at the public side of things, as this is what will be most familiar for blockchain developers.
```

To define that a variable is public, it is wrapped in the `PublicState` struct, as defined below:

#include_code public_state_struct /yarn-project/aztec-nr/aztec/src/state_vars/public_state.nr rust

The `PublicState` struct is generic over the variable type `T` and its serialized size `T_SERIALISED_LEN`. 
:::info
Currently, the length of the types must be specified when declaring the storage struct but the intention is that this will be inferred in the future.
:::

> The struct contains a `storage_slot` which similarly to Ethereum is used to figure out *where* in the storage the variable is stored. Notice that while we don't have the exact same [state model](./../../../concepts/foundation/state_model.md) as EVM chains it will from the developers points of view work very similar.

Reviewer: ```suggestion
The struct contains a `storage_slot` which, similar to Ethereum, is used to figure out *where* in storage the variable is located. Notice that while we don't have the exact same [state model](./../../../concepts/foundation/state_model.md) as EVM chains it will look similar from the contract developers point of view.
```

Beyond the struct, the `PublicState` also contains `serialization_methods`, which is a struct with methods that instruct the `PublicState` how to serialize and deserialize the variable.

#include_code TypeSerialisationInterface /yarn-project/aztec-nr/aztec/src/types/type_serialisation.nr rust

:::info

> The British haven't surrendered fully to US spelling, so serialization is currently inconsistent.

Reviewer: :joy:

```solidity
address internal admin;
```
:::info
We know its verbose, and are working on making it less so.
:::

#### Mapping example
Say we want to have a group of `minters` that are able to mint assets in our contract, and we want them in public storage, because [access control in private is quite cumbersome](./../../../concepts/foundation/communication/public_private_calls.md#a-note-on-l2-access-control). In the `Storage` struct we can add it as follows:

#include_code storage_minters /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

And then when initializing it in the `Storage::init` function we can do it as follows:

#include_code storage_minters_init /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

In this case, specifying that we are dealing with a map of Fields, and that it should be put at slot 2.

This would be similar to the following in solidity:
```solidity
mapping(address => uint256) internal minters;
```

### `read`

Now we have an idea of how to define storage, but storage is not really useful before we start using it, so how can we access it? 

> Reading data from storage is somewhat straightforward, on the `PublicState` structs we got a `read` method to read the value at the location in storage and using the specified deserialization method to deserialize it.

Reviewer: ```suggestion
Reading data from storage is straightforward. On the `PublicState` structs we have a `read` method to read the value at the location in storage and using the specified deserialization method to deserialize it.
```

- [Singleton<NoteType\>](#singletonnotetype)
- [ImmutableSingleton<NoteType\>](#immutablesingletonnotetype)
- [Set<NoteType\>](#setnotetype)

These three structs abstract-away many of Aztec's protocol complexities, by providing intuitive methods to modify notes in the utxo tree in a privacy-preserving way.

:::info
Note that an app can also choose to emit data via unencrypted log, or to define a note whose data is easy to figure out, then the information is technically not private and could be visible to anyone.
:::

### Notes 
Unlike public state variables, which can be arbitrary types, private state variables operate on `NoteType`.

Notes are the fundamental elements in the private world.

A note should conform to the following interface:

#include_code NoteInterface /yarn-project/aztec-nr/aztec/src/note/note_interface.nr rust

The interplay between a private state variable and its notes can be confusing. Here's a summary to aid intuition:

- A private state variable (of type `Singleton`, `ImmutableSingleton` or `Set`) may be declared in [Storage](storage).

> - Every note contains (as a 'header') the contract address and storage slot of the state variable to which it "belongs". A note is said to "belong" to a private state if the storage slot of the private state matches the storage slot contained in the note's header.

Reviewer: Do you think it would be worth elaborating on what's in the note header?

Author: Probably a good idea. Will elaborate more.

  - To modify the "current value" of a `Set` state variable, is to [`insert`](#insert) new notes into the `Set`, or [`remove`](#remove) notes from that set.
  - Interestingly, if a developer requires a private state to be modifiable by users who _aren't_ privy to the value of that state, a `Set` is a very useful type. The `insert` method allows new notes to be added to the `Set` without knowing any of the other notes in the set! (Like posting an envelope into a post box, you don't know what else is in there!).

## `Singleton<NoteType>`

Singleton is a private state variable that is unique in a way. When a Singleton is initialized, a note is created to represent its value. And the way to update the value is to destroy the current note, and create a new one with the updated value.

Like for public state, we define the struct to have context, a storage slot and `note_interface` specifying how the note should be constructed and manipulated.

#include_code struct /yarn-project/aztec-nr/aztec/src/state_vars/singleton.nr rust

> An example of singleton usage in in the account contracts keeping track of public keys. The `Singleton` is added to the `Storage` struct as follows:

Reviewer: ```suggestion
An example of singleton usage in the account contracts is keeping track of public keys. The `Singleton` is added to the `Storage` struct as follows:
```

Reviewer: Also, should you link to the Account contract implementation in the next line instead of the docs_example_contract, since it's explicitly mentioned.

Author: Yep, will address.

#include_code storage-singleton-declaration /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/main.nr rust

### `new`

#include_code new /yarn-project/aztec-nr/aztec/src/state_vars/singleton.nr rust

As part of the initialization of the `Storage` struct, the `Singleton` is created as follows, here at storage slot 1 and with the `NoteInterface` for `CardNote`.

#include_code start_vars_singleton /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/main.nr rust

### `initialise`

> As mention, the singleton is initialized to create the first note and value. 

Reviewer: ```suggestion
As mention, the singleton is initialized to create the first note and value. Here is the source code for the initialise function:
```

#include_code initialise /yarn-project/aztec-nr/aztec/src/state_vars/singleton.nr rust

When this function is called, a nullifier of the storage slot is created, preventing this Singleton from being initialized again.

Unlike public states, which have a default initial value of `0` (or many zeros, in the case of a struct, array or map), a private state (of type `Singleton`, `ImmutableSingleton` or `Set`) does not have a default initial value. The `initialise` method (or `insert`, in the case of a `Set`) must be called.

:::info

> Extend on what happens if you try to use non-initialized state.

Reviewer: Is this a note to self?

Author: Partially to self or someone walking through to elaborate. I haven't tested the effects currently, but should have time to do it today and add the results in here.

A commitment from the note will be generated, and inserted into data-tree, allowing us to later use in contract interactions. Recall that the content of the note should be shared with the owner to allow them to use it, as mentioned this can be done via an [encrypted log](./events.md#encrypted-events), or offchain via web2, or completely offline.

#include_code insert /yarn-project/aztec-nr/easy-private-state/src/easy_private_state.nr rust

### `insert_from_public`

While we don't support private functions to directly alter public storage, the opposite direction is possible. The `insert_from_public` allow public function to insert notes into private storage. This is very useful when we want to support private function calls that must have been initiated in public, such as shielding or the like.

#include_code insert_from_public /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust

The usage is rather straight-forward and very similar to using the `insert` method with the difference that this one is called in public functions.

#include_code insert_from_public /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

### `assert_contains_and_remove`

> This function is used to check existence of a note and then remove it without having read the note ahead of time. This can be useful for cases where the user would is providing all the information needed, such as cases where the note was never emitted to the network and thereby available to the wallet.

Reviewer: ```suggestion
This function is used to check existence of a note and then remove it without having read the note ahead of time. This can be useful for cases where the user is providing all the information needed, such as cases where the note was never emitted to the network and thereby available to the wallet.
```

Allow insertions into the rpc database and this one can be removed.
-->

### `assert_contains_and_remove_publicly_created`

Like above, this is used to ensure that the message exists in the data tree and then consume it. However, it differs slightly since there is currently a difference between notes that have been inserted from public and private execution. This means that you currently must use this function to consume and nullify a note that was created in a public function. This will be fixed in the future. 

#include_code assert_contains_and_remove_publicly_created /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust

While this might look intimidating, the use of the function is rather easy, and is used in the following way:

#include_code assert_contains_and_remove_publicly_created /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

The reason we are not reading this note ahead of time is that no [encrypted log](./events.md#encrypted-events) was emitted for this note, since it was created in public thereby making the encrypted log useless (everyone saw the content ahead of time).

### `remove`

> Will remove a note from the set if it previously have been read from storage. That being that you have fetched it through a `get_notes` call. This is useful when you want to remove a note that you have previously read from storage to not have to read it again. If you recall from earlier, we are emitting a nullifier when reading values to make sure that they are up to date. 

Reviewer: ```suggestion
Will remove a note from the set if it previously has been read from storage, e.g. you have fetched it through a `get_notes` call. This is useful when you want to remove a note that you have previously read from storage and do not have to read it again. If you recall from earlier, we are emitting a nullifier when reading values to make sure that they are up to date. 
```

#include_code remove /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust

> An example of how to use this operation is visible in the `easy_private_state`:

Reviewer: Do we have an explanation or example anywhere in the docs about why we have easy private state or how to use it? Might be good to link to it, or create a resource if we don't.

Author: Not sure there is a separate example, but generally it was to make it "easier" to work to use than sets directly such that they from dev perspective behaves kinda like a singular value. Should probably be refactored as well, will create a separate issue for this.

#include_code remove /yarn-project/aztec-nr/easy-private-state/src/easy_private_state.nr rust

### `get_notes`

This function returns the notes the account has access to:

#include_code get_notes /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust

There's a limit on the maximum number of notes this function can return at a time. Check [here](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/aztec-nr/aztec/src/constants_gen.nr) and look for `MAX_READ_REQUESTS_PER_CALL` for the up-to-date number.

> Because of this limit, we should always consider using the second argument `NoteGetterOptions` to target the notes we need, and to reduce the time required to recursively call this function.

Reviewer: What does it mean to "to reduce the time required to recursively call this function."? Can you say a bit more about why this is important?

Author: Will extend. 

The TL;DR:
When you are reading, you spend read requests for every call that you need to constrain. So the idea is that you find your notes, and then you apply a filter to those, such that only a subset is require to be read -> minimizing the number of calls you need to make and thereby the "time required to recursive call this function".

As an example, say that you have a lot of notes for a specific asset. Say you got 20 notes or something of different size. You want to transfer 10 tokens, which you can do with just a few of your tokens. You then use the filter to "stop" making more reads as soon as you found enough to perform the transfer. This could mean that you need to do just a few reads instead of reading all 20 notes, which could be the difference between using 1 second in proving or 10 seconds.

An example of this is is using the [filter_notes_min_sum](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/aztec-nr/value-note/src/filter.nr) to get "enough" notes to cover a given value. 

#include_code get_notes /yarn-project/aztec-nr/easy-private-state/src/easy_private_state.nr rust

> ### `view_notes`

Reviewer: Do we use this function anywhere in our source code? I think it would be helpful to see an example.

Author: The balance utilities is actually calling this under the hood, will add a reference to it as showcase.

Functionally similar to [`get_notes`](#get_notes), but executed unconstrained and can be used by the wallet to fetch notes for use by front-ends etc.

#include_code view_notes /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust

There's also a limit on the maximum number of notes that can be returned in one go. To find the current limit, refer to [this file](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/aztec-nr/aztec/src/constants_gen.nr) and look for `MAX_NOTES_PER_PAGE`.

The key distinction is that this method is unconstrained. It does not perform a check to verify if the notes actually exist, which is something the [`get_notes`](#get_notes) method does under the hood. Therefore, it should only be used in an unconstrained contract function.

This function requires a `NoteViewerOptions`:

#include_code NoteViewerOptions /yarn-project/aztec-nr/aztec/src/note/note_viewer_options.nr rust

The `NoteViewerOptions` is essentially similar to the [`NoteGetterOptions`](#notegetteroptions), except that it doesn't take a custom filter.

### NoteGetterOptions

> `NoteGetterOptions` encapsulates a set of configurable options for filtering and retrieving a selection of notes from a data oracle. Developers can design instances of `NoteGetterOptions`, to determine how notes should be filtered and returned to the functions of their smart contracts.

Reviewer: Can you say a bit more or link to more info about data oracles?

Author: Data oracles is essentially one of the oracles from the functions section, but can add a link in here

The indices are: 0 for `points`, 1 for `secret`, and 2 for `owner`.

In the example, `.select(2, account_address)` matches the 2nd field of `CardNote`, which is `owner`, and returns the cards whose `owner` field equals `account_address`.

`.sort(0, SortOrder.DESC)` sorts the 0th field of `CardNote`, which is `points`, in descending order.

There can be as many conditions as the number of fields a note type has. The following example finds cards whose fields match the three given values:

#include_code state_vars-NoteGetterOptionsMultiSelects /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/options.nr rust

While `selects` lets us find notes with specific values, `filter` lets us find notes in a more dynamic way. The function below picks the cards whose points are at least `min_points`:

#include_code state_vars-OptionFilter /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/options.nr rust

We can use it as a filter to further reduce the number of the final notes:

#include_code state_vars-NoteGetterOptionsFilter /yarn-project/noir-contracts/src/contracts/docs_example_contract/src/options.nr rust

> One thing to remember is, `filter` will be applied on the notes after they are picked from the database. Therefor, it's possible that the actual notes we end up getting are fewer than the limit.

Reviewer: ```suggestion
One thing to remember is, `filter` will be applied on the notes after they are picked from the database. Therefore, it's possible that the actual notes we end up getting are fewer than the limit.
```

Like reading, it is actually quite straight-forward. We have a `write` method on the `PublicState` struct that takes the value to write as an input and saves this in storage. It uses the serialization method defined earlier to serialize the value which inserts (possibly multiple) values into storage.

#include_code public_state_struct_write /yarn-project/aztec-nr/aztec/src/state_vars/public_state.nr rust

#### Writing to our `admin` example

#include_code write_admin /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

#### Writing to our `minters` example

#include_code write_minter /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

## Private State Variables
In contrast to public state, private state is persistent state that is **not** visible to the whole world. Depending on the logic of the smart contract, a private state variable's current value will only be known to one entity, or a closed group of entities.

The value of a private state variable can either be shared via an [encrypted log](./events.md#encrypted-events), or offchain via web2, or completely offline: it's up to the app developer.

> <!-- TODO: insert link in italics above -->

Reviewer 2: Can remove this and the TODO below.

#include_code insert_from_public /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust

The usage is rather straight-forward and very similar to using the `insert` method with the difference that this one is called in public functions.

#include_code insert_from_public /yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr rust

### `assert_contains_and_remove`

This function is used to check existence of a note and then remove it without having read the note ahead of time. This can be useful for cases where the user is providing all the information needed, such as cases where the note was never emitted to the network and thereby available to the wallet.

#include_code assert_contains_and_remove /yarn-project/aztec-nr/aztec/src/state_vars/set.nr rust

<!---
@LHerskind
I don't see why this one is actually needed and could not be deprecated down the line. 

> Allow insertions into the rpc database and this one can be removed.

Reviewer 2: Good suggestion! I'm working on removing it now :)
